### PR TITLE
Resident Directory Table and Pagination

### LINKED ISSUE #90: Resident Directory Table and Pagination
## Summary

The Resident Directory needs to include a table to allow users to view resident information in the system. Here are the columns we'd want to show in the table:
- Resident ID- This should be a combination of the resident's initials and their room_num (ex: SC420)
- Date Joined- When the resident entered the system
- Date Left- When the resident left the system, can be null
- Building- The building the resident lives in

In this ticket, I'd also like the back-end/front-end integration to be written so that the data from the database is shown on the table. One issue is that the original `GET` method in `resident_routes` needs to be tweaked for pagination. Refer to how pagination is handled on the `log_record_routes` `GET` method. #102 is also relevant here since whatever improvements we make to pagination should also go here. 

This also means pagination should exist at the bottom of the page. There is a component for Pagination called `Pagination.tsx`, but it only supports log record data, and it should be tweaked to support residents (and employees as well). 

## Goals

- [ ] Create the front-end components to display resident data in a tabular format
- [ ] Write the API calls to display resident data from the DB in the table
- [ ] Add Pagination at the bottom to allow the user to view the results in separate pages

## Implementation

1. Start by creating a table using Chakra, and render some mock data. 
2. From there, write the code for the API calls to render data from the DB.
3. Tweak the API so that Pagination is considered.
4. Tweak the implementation of the pagination component so it supports resident data.
5. Add the pagination component to the page and validate it works correctly.

### PR DESCRIPTION
**Pull request by kevin-pierce**
## GitHub Issue link

[Resident Directory](https://github.com/uwblueprint/supportive-housing/issues/90)

## Implementation description
Screen Shot 2023-07-12 at 9 09 59 AM

- Implemented table component for Resident directory page
- Moved getResidents to `ResidentAPIClient`
- Updated params of getResidents service, route, and API client function
- Hooked up pagination component to resident directory table

## Steps to test
1. Create test residents in your local DB
2. Verify that test residents appear in table + pagination works as intended
3. Ensure that residents with no end date in the DB have the "Current" status in the table
4. Ensure that residents WITH an end date in the DB have the "Past" status in the table

## What should reviewers focus on?
- Column widths feel a bit wonky ; I adjusted the widths adhoc, but as per discussion with @Safewaan this is something we'll revisit later 
- Didn't implement the "Add Resident" button, but rather marked it as a TODO in with a comment ; we can add this back in if needed

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR
- [ ] If I have made API changes, I have updated the [REST API Docs](https://www.notion.so/uwblueprintexecs/REST-Endpoints-05ce60312bb943439dfda42bb1318536)
- [ ] IF I have made changes to the db/models, I have updated the [Data Models Page](https://www.notion.so/uwblueprintexecs/Data-Models-760f8aa06b244eb0842c079ad77987b0)
- [ ] I have updated other Docs as needed

### CODE REVIEW

File: backend/app/services/implementations/residents_service.py [hunk 1]
```diff
@@ -82,13 +82,29 @@ def delete_resident(self, resident_id):
             )
         db.session.commit()
 
-    def get_residents(self, resident_id=None):
+    def get_residents(self, return_all, page_number, results_per_page=10, resident_id=None):
```
Comment (safchow) [inline backend/app/services/implementations/residents_service.py]: I think the pagination improvements Connor and Owen made are missing here... Connor has them in his PR: #135 using the ORM, and in log_records, it's done using raw sql. @connor-bechthold can you take a look with Kevin here and see where the changes should be made?

File: frontend/src/theme/common/logsTableStyles.tsx [hunk 2]
```diff
@@ -3,7 +3,7 @@ import fontStyles from "../fontStyles";
 
 const Table: ComponentStyleConfig = {
   variants: {
-    logRecordsTable: {
+    baseTable: {
```
Comment (safchow) [inline frontend/src/theme/common/logsTableStyles.tsx]: think there might be some merge conflicts with this and #135, this is named showTable in that PR

File: backend/app/services/implementations/residents_service.py (hunk 1 above)
Comment (safchow) [reply backend/app/services/implementations/residents_service.py]: also I think there's a separate api to get the count now, sync with Connor to figure out what needs to be changed 🐸

File: frontend/src/components/pages/ResidentDirectory/ResidentDirectoryTable.tsx [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+              <Th>Residency End Date</Th>
+            </Tr>
+          </Thead>
+
+          <Tbody>
+            {residents.map((resident) => {
+              const residentId = resident.initial + resident.roomNum;
+              const { startDate, endDate, status } = getFormattedDatesAndStatus(resident);
+                
+              return (
```
Comment (safchow) [inline frontend/src/components/pages/ResidentDirectory/ResidentDirectoryTable.tsx:68]: noticing some weird behaviour in the table when there's only one item:
image

Seems to occur in log records as well, if you can find a fix for this (I think its just setting a max-height on each table row) awesome, if not I can make this a bug fix ticket

Comment (safchow) [reply frontend/src/components/pages/ResidentDirectory/ResidentDirectoryTable.tsx:68]: with two residents:
image

**Review by safchow — CHANGES_REQUESTED** [review]
everything is working correctly, amazing job 🚀 👍 😂 🐮 🌴 

One thing that needs to be changed is the pagination- Connor and Owen worked on some improvements for it and I think they should go in here as well for consistency. I don't think we'll ever have more than 10000 residents so there's not going to be a notable difference in performance but we should keep our code consistent, it might be best to sync with Connor for these changes

File: frontend/src/theme/common/logsTableStyles.tsx (hunk 2 above)
Comment (kevin-pierce) [reply frontend/src/theme/common/logsTableStyles.tsx]: Resolved

File: backend/app/rest/residents_routes.py [hunk 4]
```diff
@@ -104,10 +104,44 @@ def get_residents():
...
+    except:
+        pass
+
+    results_per_page = 10
+    try:
+        results_per_page = int(request.args.get("results_per_page"))
+    except:
+        pass
+
+    try:
```
Comment (kevin-pierce) [inline backend/app/rest/residents_routes.py]: We should revisit this at some point 😢

File: backend/app/services/implementations/residents_service.py (hunk 1 above)
Comment (kevin-pierce) [reply backend/app/services/implementations/residents_service.py]: Resolved

File: backend/app/rest/residents_routes.py (hunk 4 above)
Comment (connor-bechthold) [reply backend/app/rest/residents_routes.py]: Disagree this is BEAUTIFUL

File: backend/app/services/implementations/residents_service.py [hunk 5]
```diff
@@ -82,13 +82,29 @@ def delete_resident(self, resident_id):
...
+                residents_results = (
+                    Residents.query.limit(results_per_page)
+                    .offset((page_number - 1) * results_per_page)
+                    .all()
+                )
+                
+            residents_results = list(map(lambda resident: resident.to_dict(), residents_results))
+            return {"residents": residents_results}
 
-            return self.to_json_list(residents_results)
```
Comment (connor-bechthold) [inline backend/app/services/implementations/residents_service.py]: We should be able to delete this now

File: backend/app/services/implementations/residents_service.py (hunk 1 above)
Comment (connor-bechthold) [inline backend/app/services/implementations/residents_service.py]: Nit: Don't need this defaulted to 10 if it's already being defaulted to 10 in the route. Can probably leave this though

File: frontend/src/components/pages/ResidentDirectory/ResidentDirectory.tsx [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+      <Box
+        textStyle="dm-sans-font"
+        textAlign="center"
+        width="75%"
+        paddingTop="2%"
+        margin="0px auto"
+        color="blue.600"
+      >
+        <Flex marginBottom="16px">
+          <Box textStyle="header-large">Resident Directory</Box>
```
Comment (connor-bechthold) [inline frontend/src/components/pages/ResidentDirectory/ResidentDirectory.tsx]: Should be using `hero-table` for this (styling might looks broken but if you merge main it should be fine)

**Review by safchow — APPROVED** [review]
DEMON PR 💪 amazing work yet again 

also thanks for making the pagination changes :')

File: frontend/src/components/pages/ResidentDirectory/ResidentDirectoryTable.tsx [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+              <Th>Status</Th>
+              <Th>Building</Th>
+              <Th>Residency Start Date</Th>
+              <Th>Residency End Date</Th>
+            </Tr>
+          </Thead>
+
+          <Tbody>
+            {residents.map((resident) => {
+              const residentId = resident.initial + resident.roomNum;
```
Comment (connor-bechthold) [inline frontend/src/components/pages/ResidentDirectory/ResidentDirectoryTable.tsx]: Hm I thought this was apart of the resident table? cc @Safewaan

**Review by connor-bechthold — CHANGES_REQUESTED** [review]

File: frontend/src/components/pages/ResidentDirectory/ResidentDirectoryTable.tsx (hunk 7 above)
Comment (safchow) [reply frontend/src/components/pages/ResidentDirectory/ResidentDirectoryTable.tsx]: yeah it is, I merged my fix so rebase main and update this

Comment (safchow) [reply frontend/src/components/pages/ResidentDirectory/ResidentDirectoryTable.tsx]: @kevin-pierce tl;dr get residents pulls the resident id so you don't have to do this step

File: frontend/src/components/pages/ResidentDirectory/ResidentDirectoryTable.tsx (hunk 3 above)
Comment (kevin-pierce) [reply frontend/src/components/pages/ResidentDirectory/ResidentDirectoryTable.tsx:68]: Resolved in this PR: https://github.com/uwblueprint/supportive-housing/pull/147

File: backend/app/services/implementations/residents_service.py (hunk 5 above)
Comment (kevin-pierce) [reply backend/app/services/implementations/residents_service.py]: ✂️

File: backend/app/services/implementations/residents_service.py (hunk 1 above)
Comment (kevin-pierce) [reply backend/app/services/implementations/residents_service.py]: Good point ; removed

File: frontend/src/components/pages/ResidentDirectory/ResidentDirectoryTable.tsx (hunk 7 above)
Comment (kevin-pierce) [reply frontend/src/components/pages/ResidentDirectory/ResidentDirectoryTable.tsx]: Resolved ; just updating the type as well in this PR

Comment (kevin-pierce) [reply frontend/src/components/pages/ResidentDirectory/ResidentDirectoryTable.tsx]: So there seems to be a bit of discrepancy between the `Resident` type we use for creation + fetching ; the residents we fetch from our DB have the `residentId` field, but the type involved in `addResident` is different in that it _does not need to supply this field__  ; as a temporary solution to expedite this work, I've just used an non-null assertion on the `residentId` field, and I'll resolve the type issues separately

Comment (kevin-pierce) [reply frontend/src/components/pages/ResidentDirectory/ResidentDirectoryTable.tsx]: Created - https://github.com/uwblueprint/supportive-housing/issues/148 to track this

**Review by connor-bechthold — APPROVED** [review]
WICKED 👨‍🍳

**Pull request merged**